source/opt/ilist.h

```diff
@@ -0,0 +1,194 @@
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_ILIST_H_
+#define LIBSPIRV_OPT_ILIST_H_
+
+#include <cassert>
+#include <type_traits>
+
+#include "ilist_node.h"
+
+namespace spvtools {
+namespace utils {
+
+// An IntrusiveList is a generic implementation of a doubly-linked list.  The
+// intended convention for using this container is:
+//
+//      class Node : public IntrusiveNodeBase<Node> {
+//        // Node that "Node", the class being defined is the template.
+//        // Must have a default constructor accessible to List.
+//        // Add whatever data is needed in the node
+//      };
+//
+//      typedef IntrusiveList<Node> List;
+//
+// You can also inherit from IntrusiveList instead of a typedef if you want to
+// add more functionality.
+//
+// The condition on the template for IntrusiveNodeBase is there to add some type
+// checking to the container.  The compiler will still allow inserting elements
+// of type IntrusiveNodeBase<Node>, but that would be an error. This assumption
+// allows NextNode and PreviousNode to return pointers to Node, and casting will
+// not be required by the user.
+
+template <class NodeType>
+class IntrusiveList {
+ public:
+  static_assert(
+      std::is_base_of<IntrusiveNodeBase<NodeType>, NodeType>::value,
+      "The type from the node must be derived from IntrusiveNodeBase, with "
+      "itself in the template.");
+
+  // Creates an empty list.
+  inline IntrusiveList();
+
+  // Moves the contents of the given list to the list being constructed.
+  IntrusiveList(IntrusiveList&&);
+
+  // Destorys the list.  Note that the elements of the list will not be deleted,
+  // but thy will be removed from the list.
+  ~IntrusiveList();
+
+  // Moves all of the elements in the list on the RHS to the list on the LHS.
+  IntrusiveList& operator=(IntrusiveList&&);
+
+  // Basetype for iterators so an IntrusiveList can be traversed like STL
+  // containers.
+  template <class T>
+  class iterator_template {
+   public:
+    inline iterator_template(const iterator_template& i) : node_(i.node_) {}
+    inline iterator_template& operator++() {
+      node_ = node_->NextNode();
+      return *this;
+    }
+    inline iterator_template& operator--() {
+      node_ = node_->PreviousNode();
+      return *this;
+    }
+    inline iterator_template& operator=(const iterator_template& i) {
+      node_ = i.node_;
+      return *this;
+    }
+    inline T& operator*() const { return *node_; }
+    inline T* operator->() const { return node_; }
+
+    friend inline bool operator==(const iterator_template& lhs,
+                                  const iterator_template& rhs) {
+      return lhs.node_ == rhs.node_;
+    }
+    friend inline bool operator!=(const iterator_template& lhs,
+                                  const iterator_template& rhs) {
+      return !(lhs == rhs);
+    }
+
+   private:
+    inline iterator_template(T* node) { node_ = node; }
```

Comment:
Consider defining or explicitly disabling the default constructor.

Reply:
Done